Add MemberDowngrade failpoint

siyuanfoundation opened the pull request:
Please read https://github.com/etcd-io/etcd/blob/main/CONTRIBUTING.md#contribution-flow.

https://github.com/etcd-io/etcd/issues/17118

siyuanfoundation commented:
I am getting the following error sometimes. @serathius do you know what this error usually comes from?
```
Error:      	Received unexpected error:
        	            	failed to read WAL, cannot be repaired, err: wal: slice bounds out of range
```

serathius commented on tests/robustness/failpoint/cluster.go:
```diff
@@ -138,6 +142,92 @@ func (f memberReplace) Available(config e2e.EtcdProcessClusterConfig, member e2e
+	}
+	v3_6 := semver.Version{Major: 3, Minor: 6}
+	// only current version cluster can be downgraded.
+	return config.ClusterSize > 1 && v.Compare(v3_6) >= 0 && (config.Version == e2e.CurrentVersion && member.Config().ExecPath == e2e.BinPath.Etcd)
```
Why only cluster size > 1?

ahrtr replied:
+1 the same question.

serathius commented on tests/robustness/failpoint/cluster.go:
```diff
@@ -138,6 +142,92 @@ func (f memberReplace) Available(config e2e.EtcdProcessClusterConfig, member e2e
+		return nil, err
+	}
+	targetVersion := semver.Version{Major: v.Major, Minor: v.Minor - 1}
+	numberOfMembersToDowngrade := rand.Int()%len(clus.Procs) + 1
```
Why not downgrade all members?

ahrtr commented:
Overall looks good to me, please mark this PR as ready to review when you feel comfortable.

siyuanfoundation commented:
> > I am getting the following error sometimes. @serathius do you know what this error usually comes from?
> > ```
> > Error:      	Received unexpected error:
> >         	            	failed to read WAL, cannot be repaired, err: wal: slice bounds out of range
> > ```
> 
> Most likely the WAL record is somehow corrupted again. Added more debug info in #19067.

Updated the error message.

ahrtr commented:
> Updated the error message. I get this error almost always when I run the robustness test for 20 times. Does that mean the downgrade would likely corrupt WAL?
> 
> I tried `member.Stop()` instead of `member.Kill()` and did not change the results.

Did not get time to dig into the robustness test's use case of reading WAL file, but my immediate feeling is the reason might be due to an inappropriate snapshot {Index: 0, Term: 0} is always set for the WAL.  

https://github.com/etcd-io/etcd/blob/e0bbea994feb7638d966a413a8934c477bd784a7/tests/robustness/report/wal.go#L110

If the v2 snapshot files have already been rotated, in other words, the very first snapshot files have been purged, then it means there are data loss (you are not reading WAL records right following the v2 snapshot Index) from v2store perspective (although there isn't data loss from v3store perspective), then you will definitely see this error.
https://github.com/etcd-io/etcd/blob/e0bbea994feb7638d966a413a8934c477bd784a7/server/storage/wal/wal.go#L488-L493

A thought to double check this... set a big value for both `--max-snapshots` and `--max-wals` to ensure no snapshot files and WAL files rotation happens.

siyuanfoundation commented:
@ahrtr Thanks for the suggestion.
I tried `"--max-wals=1000", "--max-snapshots=1000"`, still seeing the same failure. So it is probably not due to rotation?

ahrtr commented:
> @ahrtr Thanks for the suggestion. I tried `"--max-wals=1000", "--max-snapshots=1000"`, still seeing the same failure. So it is probably not due to rotation?

The WAL files should haven't rotated, otherwise it will fail to find the WAL files which match the snap index.

https://github.com/etcd-io/etcd/blob/0966b4d68ec9ff443a43431eae71305796e3200f/server/storage/wal/wal.go#L405-L408 

Please use the `etcd-dump-logs` to analyse the WAL files, to double check whether the WAL records are incremental by 1.  Run command below,

```
./etcd-dump-logs --start-index=0 path/2/data-dir
```

Also I tried to test it locally, but couldn't reproduce it. Please provide detailed step, I may try it when I get bandwidth.

```
go test -run TestRobustnessExploratory -v --count 100 --failfast --timeout 1h
```

ahrtr commented:
> Please use the `etcd-dump-logs` to analyse the WAL files, to double check whether the WAL records are incremental by 1. Run command below,
> 
> ```
> ./etcd-dump-logs --start-index=0 path/2/data-dir
> ```

Refer to
https://github.com/etcd-io/etcd/pull/19082

siyuanfoundation commented:
I added my local patch to force test this failpoint only. 
``` bash
# download 3.5 release with 
make test-e2e-release

make gofail-enable && make build
cd tests/robustness
go test -run TestRobustnessExploratory -v --count 100 --failfast --timeout 1h
```

siyuanfoundation commented:
The error msg is:
```
wal: slice bounds out of range, snapshot[Index: 0, Term: 0], current entry[Index: 12487, Term: 2], len(ents): 11243
```
and the log dump is

```
# member-2
Snapshot:
term=2 index=12486 nodes=[ac4ec652f10e5b49 bf19ae4419db00dc eabdbb777cf498cb] confstate={"voters":[12416079282240904009,13770228943176794332,16914881897345358027],"auto_leave":false}
Start dumping log entries from snapshot.
WAL metadata:
nodeID=bf19ae4419db00dc clusterID=b3bc0c1919fe5d7e term=3 commitIndex=36822 vote=eabdbb777cf498cb
WAL entries: 24337
lastIndex=36823
term         index      type    data
   2         12487      norm    header:<ID:11010056292433083151 > txn:<compare:<target:MOD key:"/registry/pods/default/hyG1G" mod_revision:8853 > success:<request_put:<key:"/registry/pods/default/hyG1G" value:"11523" > > failure:<request_range:<key:"/registry/pods/default/hyG1G" > > > 
   2         12488      norm    header:<ID:6577951309147099754 > txn:<compare:<target:MOD key:"/registry/pods/default/VIap8" mod_revision:8868 > success:<request_put:<key:"/registry/pods/default/VIap8" value:"11525" > > failure:<request_range:<key:"/registry/pods/default/VIap8" > > > 

```

```
# member-0
Snapshot:
term=2 index=8003 nodes=[ac4ec652f10e5b49 bf19ae4419db00dc eabdbb777cf498cb] confstate={"voters":[12416079282240904009,13770228943176794332,16914881897345358027],"auto_leave":false}
Start dumping log entries from snapshot.
WAL metadata:
nodeID=eabdbb777cf498cb clusterID=b3bc0c1919fe5d7e term=3 commitIndex=36822 vote=eabdbb777cf498cb
WAL entries: 28820
lastIndex=36823
term         index      type    data
   2          8004      norm    header:<ID:11010056292433081493 > txn:<compare:<target:MOD key:"/registry/pods/default/bRqG1" mod_revision:5736 > success:<request_put:<key:"/registry/pods/default/bRqG1" value:"7367" > > failure:<request_range:<key:"/registry/pods/default/bRqG1" > > > 
   2          8005      norm    header:<ID:11010056292433081494 > txn:<compare:<target:MOD key:"/registry/pods/default/3ApAH" mod_revision:5698 > success:<request_delete_range:<key:"/registry/pods/default/3ApAH" > > failure:<request_range:<key:"/registry/pods/default/3ApAH" > > > 

... # all the entries in between
2         12484      norm    header:<ID:11010056292433083149 > txn:<compare:<target:MOD key:"/registry/pods/default/7i4E8" mod_revision:8852 > success:<request_put:
<key:"/registry/pods/default/7i4E8" value:"11521" > > failure:<request_range:<key:"/registry/pods/default/7i4E8" > > > 
   2         12485      norm    header:<ID:11010056292433083150 > txn:<compare:<target:MOD key:"/registry/pods/default/UonNv" mod_revision:8829 > success:<request_put:
<key:"/registry/pods/default/UonNv" value:"11522" > > failure:<request_range:<key:"/registry/pods/default/UonNv" > > > 
   2         12486      norm    header:<ID:6577951309147099753 > txn:<compare:<target:MOD key:"/registry/pods/default/iDLqp" mod_revision:8839 > success:<request_delet
e_range:<key:"/registry/pods/default/iDLqp" > > failure:<request_range:<key:"/registry/pods/default/iDLqp" > > > 
   2         12487      norm    header:<ID:11010056292433083151 > txn:<compare:<target:MOD key:"/registry/pods/default/hyG1G" mod_revision:8853 > success:<request_put:<key:"/registry/pods/default/hyG1G" value:"11523" > > failure:<request_range:<key:"/registry/pods/default/hyG1G" > > > 
   2         12488      norm    header:<ID:6577951309147099754 > txn:<compare:<target:MOD key:"/registry/pods/default/VIap8" mod_revision:8868 > success:<request_put:<key:"/registry/pods/default/VIap8" value:"11525" > > failure:<request_range:<key:"/registry/pods/default/VIap8" > > > 
   2         12489      norm    header:<ID:6577951309147099755 > txn:<compare:<target:MOD key:"/registry/pods/default/ZYXYI" mod_revision:8865 > success:<request_put:<key:"/registry/pods/default/ZYXYI" value:"11526" > > failure:<request_range:<key:"/registry/pods/default/ZYXYI" > > > 
   2         12490      norm    method=PUT path="/0/members/bf19ae4419db00dc/attributes" val="{\"name\":\"TestRobustnessExploratoryKubernetesHighTrafficClusterOfSize3-test-2\",\"clientURLs\":[\"http://localhost:20010\"]}"

```

ahrtr commented:
I reproduced this locally.

```
failed to read WAL, cannot be repaired, err: wal: slice bounds out of range, snapshot[Index: 0, Term: 0], current entry[Index: 11282, Term: 2], len(ents): 10522
```

## Initial analysis
The direct reason is that some WAL entries were somehow missing right the snapshot.

```
   2	     10520	norm	header:<ID:6577951317820541023 > txn:<compare:<target:MOD key:"/registry/pods/default/w9Kjw" mod_revision:6857 > success:<request_put:<key:"/registry/pods/default/w9Kjw" value:"9686" > > failure:<request_range:<key:"/registry/pods/default/w9Kjw" > > > 
   2	     10521	norm	header:<ID:62087081945565313 > txn:<compare:<target:MOD key:"/registry/pods/default/pGSSX" mod_revision:6868 > success:<request_put:<key:"/registry/pods/default/pGSSX" value:"9683" > > failure:<request_range:<key:"/registry/pods/default/pGSSX" > > > 
   2	     10522	norm	header:<ID:62087081945565314 > txn:<compare:<target:MOD key:"/registry/pods/default/gP0lD" mod_revision:6867 > success:<request_put:<key:"/registry/pods/default/gP0lD" value:"9684" > > failure:<request_range:<key:"/registry/pods/default/gP0lD" > > > 
   0	         0	norm	
   0	         0	norm	
   0	         0	norm
   ......
   0	         0	norm	
   2	     11282	norm	header:<ID:11010056301106530314 > txn:<compare:<target:MOD key:"/registry/pods/default/rr6ak" mod_revision:7420 > success:<request_put:<key:"/registry/pods/default/rr6ak" value:"10401" > > failure:<request_range:<key:"/registry/pods/default/rr6ak" > > > 
   2	     11283	norm	method=PUT path="/0/members/ac4ec652f10e5b49/attributes" val="{\"name\":\"TestRobustnessExploratoryKubernetesHighTrafficClusterOfSize3-test-1\",\"clientURLs\":[\"http://localhost:20005\"]}"
   2	     11284	norm	header:<ID:62087081945565542 > txn:<compare:<target:MOD key:"/registry/pods/default/zloLS" mod_revision:7438 > success:<request_delete_range:<key:"/registry/pods/default/zloLS" > > failure:<request_range:<key:"/registry/pods/default/zloLS" > > > 
```

The good news is that the data across the three members are still consistent,

```
   ubuntu@vm2:~/go/src/github.com/ahrtr/etcd$ ./bin/etcdutl hashkv /tmp/TestRobustnessExploratory_KubernetesHighTraffic_ClusterOfSize3_MemberDowngrade/1734617398669080611/server-TestRobustnessExploratoryKubernetesHighTrafficClusterOfSize3-test-0/member/snap/db -w table
+------------+---------------+------------------+
|    HASH    | HASH REVISION | COMPACT REVISION |
+------------+---------------+------------------+
| 3080225026 |         17303 |            17041 |
+------------+---------------+------------------+
ubuntu@vm2:~/go/src/github.com/ahrtr/etcd$ 
ubuntu@vm2:~/go/src/github.com/ahrtr/etcd$ ./bin/etcdutl hashkv /tmp/TestRobustnessExploratory_KubernetesHighTraffic_ClusterOfSize3_MemberDowngrade/1734617398669080611/server-TestRobustnessExploratoryKubernetesHighTrafficClusterOfSize3-test-1/member/snap/db -w table
+------------+---------------+------------------+
|    HASH    | HASH REVISION | COMPACT REVISION |
+------------+---------------+------------------+
| 3080225026 |         17303 |            17041 |
+------------+---------------+------------------+
ubuntu@vm2:~/go/src/github.com/ahrtr/etcd$ 
ubuntu@vm2:~/go/src/github.com/ahrtr/etcd$ 
ubuntu@vm2:~/go/src/github.com/ahrtr/etcd$ ./bin/etcdutl hashkv /tmp/TestRobustnessExploratory_KubernetesHighTraffic_ClusterOfSize3_MemberDowngrade/1734617398669080611/server-TestRobustnessExploratoryKubernetesHighTrafficClusterOfSize3-test-2/member/snap/db -w table
+------------+---------------+------------------+
|    HASH    | HASH REVISION | COMPACT REVISION |
+------------+---------------+------------------+
| 3080225026 |         17303 |            17041 |
+------------+---------------+------------------+
```

## Next step of the investigation

- Add verification check in the raft loop to ensure any raft Entry will never have 0 term or Index.

https://github.com/etcd-io/etcd/blob/9fa35e53f429ca8f21b0d6b26f24e1848f2652a6/server/etcdserver/raft.go#L180-L184

- Turn off lazyfs
   Is the lazyfs enabled?  If yes, then the data loss might be caused by lazyfs, but did not get time to verify it. 

cc @siyuanfoundation @fuweid @serathius 


## Proposal on how to replay WAL in robustness test

- You need to ensure the WAL files aren't rotated if you want to read the whole history, i.e. setting a big value for `--max-wals`;
  - Based on my test, the default value (5 files, and each 64MB) is good enough.
- Currently robustness test just reads all WAL records based on snap {Index:0, term:0}. Suggest to read WAL files based on snapshots. For example, you have 2 snapshot files, so read WAL records based snap 1 firstly, then continue to read WAL records based on snap 2 (assuming snap2.Index > snap1.Index).
  - So you also need to ensure the snapshot files aren't rotated.

siyuanfoundation commented:
Thanks @ahrtr for the investigation! 
This is not due to lazyfs, because lazyfs is not enabled in cluster size 3. 

@serathius do we need all the records for robustness test? Or the continuous records before the gap is enough?

serathius commented:
> @serathius do we need all the records for robustness test? Or the continuous records before the gap is enough?

Yes, currently robustness tests assume all records are persisted. Sorry, I assumed that the error came from etcd, not from robustness tests.

ahrtr commented on tests/robustness/failpoint/cluster.go:
```diff
@@ -138,6 +145,80 @@ func (f memberReplace) Available(config e2e.EtcdProcessClusterConfig, member e2e
+			return nil, err
+		}
+		member.Config().ExecPath = e2e.BinPath.EtcdLastRelease
+		err = patchArgs(member.Config().Args, "initial-cluster-state", "existing")
```
It isn't necessary to patch the flag `--initial-cluster-state`, because the data-dir and wal-dir have already been created & initialised. It won't do any harm to do it, but suggest to remove it for simplicity.

ahrtr reviewed (approved):
LGTM

Thanks @siyuanfoundation for the nice work!

I'd like to get https://github.com/etcd-io/etcd/pull/19095 merged firstly to avoid the [failed to read WAL](https://github.com/etcd-io/etcd/pull/19038#issuecomment-2532919530) flaky.

ahrtr commented:
@siyuanfoundation Please continue to work on https://github.com/etcd-io/etcd/issues/17118 when you get bandwidth, thx
